src/arch/arm/lib/debug/stack_iter.c

```diff
@@ -18,11 +18,11 @@
 #include "stack_iter.h"
 
 int stack_iter_next(stack_iter_t *f) {
-	f->fp = (void*) *((int*)f->fp - 3);
-	f->pc = (void*) *((int*)f->fp);
-	f->lr = (void*) *((int*)f->fp - 1);
+	f->pc = (void *) *((int *)f->fp);
+	f->lr = (void *) *((int *)f->fp);
+	f->fp = (void *) *((int *) (f->fp - 4));
```

Comment:
considering old and new version assumes 4 word stack frame header, it's APCS. But in APCS header is 
```
    save code pointer       [fp]        <-fp points to here
    return link value       [fp, #-4]
    return sp value         [fp, #-8]
    return fp value         [fp, #-12]
```
(https://www.cl.cam.ac.uk/~fms27/teaching/2001-02/arm-project/02-sort/apcs.txt)

Could you describe why old code is not correct, when it's apparently is?

Reply:
Basicly, new version works with `-fno-omit-frame-pointer` and old version doesn't 🤔
However, it works with `-mapcs-frame`, so ok, I will roll back these changes :)

Reply:
(by the way, old code was incorrect anyway because we need to do `f->fp = (void*) *((int*)f->fp - 0xc);`, not `f->fp = (void*) *((int*)f->fp - 3);`

Reply:
As far as I understand, both flags should be specified (`-fno-omit-frame-pointer -mapcs-frame`) to unwind stack in such way. 
Without `-fno-omit-frame-pointer` compiler is allowed to not generate full frame header for leaf functions.
Without `-mapcs-frame`, frame can't be unwinded without consulting debug info, generated by compiler (those .extab, .exidx sections)
I believe, with other options it will break from time to time

Reply:
It seems that you are right. Thanks for your remark.

Reply:
> By the way, old code was incorrect anyway because we need to do f->fp = (void*) *((int*)f->fp - 0xc);, not f->fp = (void*) *((int*)f->fp - 3);

Sorry, saw this too late. Just to be clear.

`(int*)fp - 3` (was) is the same as `(int*) (fp - 12)` (in your recent commit). But `(int*)f->fp - 0xc` from the message is not and it's not correct, it's access 12 words below `fp`.

Actually, I would prefer `(int*)fp - 3` than `(int*) (fp - 12)` because fp is `void*` and it's gcc extension to C that allows that. Maybe `uint32_t*` is better that `int*` here...